MagneticField/VolumeBasedEngine/src/MagGeometry.cc

```diff
@@ -15,14 +15,17 @@
 #include "FWCore/Utilities/interface/isFinite.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 
-#include "MagneticField/Layers/interface/MagVerbosity.h"
-#include "FWCore/MessageLogger/interface/MessageLogger.h"
-
 #include <iostream>
 
 using namespace std;
 using namespace edm;
 
+namespace {
+  std::atomic<int> instanceCounter(0);
+  thread_local int localInstance = 0;
+  thread_local MagVolume const* lastVolume = nullptr;
```

Comment:
considering thread_local is restrictive in our environment and should be avoided,
how would this differ from using e.g. ```std::map<std::thread::id, MagVolume const*> lastVolume;``` and then accessing it via ```lastVolume[std::this_thread::get_id()]```? 
the cost of navigating the map is likely still much smaller than calling the volume inTesla.
A slightly faster version could use a predetermined array size with the id conversion to int of the array size.

@makortel @VinInn @Dr15Jones

Reply:
Every access to an `std::map` shared across threads would need to be protected e.g. with a `mutex`. I would rather not add synchronization points.

Reply:
unordered_map with a large enough reserve wouldn't need a sync, or would it?

Reply:
I'd be happy to stay with thread_local, however if this is acceptable, it should be acceptable in all packages where it's reasonably needed.
Can we go for this precedent?

Reply:
In the meanwhile I finally managed test the cache re-set on several threads running in parallel (besides specifiyng process.options.numberOfThreads and numberOfStreams I also had to turn my analyzer into a stream analyzer; thanks to @VinInn for the tip). 

So I verified that the cache is re-set correctly in each of the active threads at each IOV transition.

@slava77 : I'm not the one who can answer your last question, but my 2 cents: I do understand your concern about setting a precedent. However, MF is a fundamental service for pretty much anything in SIM, RECO and HLT, not some obscure analysis package. 
If it sets a precedent, I think it also sets a pretty high standard on what is to be considered reasonable...

Regarding the (unordered) map: nice idea for what I am concerned, but I don't have enough experience to tell whether it's better than the TLS one. If experts agree it is, I can try implementing it as well.

Reply:
> unordered_map with a large enough reserve wouldn't need a sync, or would it?

It would, standard does not guarantee anything between synchronization between threads without the user doing it.

> the cost of navigating the map is likely still much smaller than calling the volume inTesla.

Maybe, but it would slow down the fast path, i.e. workloads with large cache hit ratios would have an impact.

> If it sets a precedent, I think it also sets a pretty high standard on what is to be considered reasonable...

I agree, if we are setting a precedent here, we should be really careful.

Reply:
> > the cost of navigating the map is likely still much smaller than calling the volume inTesla.
> 
> Maybe, but it would slow down the fast path, i.e. workloads with large cache hit ratios would have an impact.

Just to clarify my earlier comment, I understand that selecting a correct cache pointer will be more costly. My point is that even the fastest path that has the cache immediately, still has to call the interpolation math ("the volume inTesla"), which is likely much more expensive than any reasonable cache retrieval from non-thread_local mapping/array using thread id.

Reply:
The interpolation is quite efficient; it is reduced to one coordinate transformation + a linear interpolation within a regular grid + a RF transformation of the result. I can try timing the overhead of your suggestion in a cache-hit scenario - but I'd save the effort if the sync issue is considered to be a showstopper...

Reply:
> but I'd save the effort if the sync issue is considered to be a showstopper...

In general we should avoid adding synchronization points that the framework is unaware of (they will decrease CPU efficiency).

Reply:
can the framework provide us with a thread id -> int index map object (best if the range of int is small and can be used as an array index) that would not require synchronization on the client side?
This way the framework will be aware of how it's filled and we'll just have an array of cache pointers (possibly atomic) to read from in the MF code instead of a thread_local.

Reply:
I replied on the general "per-thread" question in https://github.com/cms-sw/cmssw/pull/28180#issuecomment-620174366.

An array with per-thread indexing would technically work, but for a good performance one has to give each thread a separate cache line to prevent false sharing (so a simple 'std::array<MagVolume const*, N>`  would not be sufficient). How would you decide the array size? Strictly speaking we do not guarantee to keep the same threads though a job. And allowing resize leads to the need to lock/synchronize (there may be ways to do an update without locking with a price of complicated code).

Reply:
for the array I was thinking (if a fixed thread indexing is not practical) to  keep it ```std::array<std::atomic<MagVolume const*>, N>``` and do ```(int)thread_index%N```
where N can be a reasonable max (like 256)

Reply:
WAIT. the framework should guarantee NOT to exceed the number of "processing" threads in the configuration. I do not think that the framework should be free to add a couple of threads to overcome a temporary shortage. If the framework wants to have such a freedom it should reserve  those thread in advance.

I proposed  somewhere above a possible implementation of all this.
Essentially the framework should provide a global function that return a struct with all info (process-status?)  that may be of interest during processing (event,setup, module etc as pointers or just identifiers) including an "index" that is per thread and per stream and it is sequential and has a known maximum (returned by another global function or stored in the struct above). Internally will use a thread_local and I suggest to locate  thise functions in the "main" program (statically linked)